### PR TITLE
Overload == and isequal for checking equality of Terms and Schemas

### PR DESCRIPTION
**Pull request by mrrobot-2000**
#173

### CODE REVIEW

File: src/terms.jl [hunk 1]
```diff
@@ -378,7 +418,7 @@ end
 
 Base.show(io::IO, t::InterceptTerm{H}) where {H} = print(io, H ? "1" : "0")
 
-Base.show(io::IO, t::ContinuousTerm) = print(io, t.sym)
+Base.show(io::IO, t::ContinuousTerm) = print(io, t.sym,"\n" , t.mean,"\n" , t.var, "\n" ,t.min,"\n" , t.max)
```
Comment (kleinschmidt) [inline src/terms.jl]: this is irrelevant for checking equality and breaks existing behavior.  It also causes the tests to fail.  As a general suggestion, make sure that the test suite runs (by doing `Pkg.test()` or running `test StatsModels` in pkg mode) before you submit a PR.

File: test/schema.jl [hunk 2]
```diff
@@ -1,10 +1,17 @@
...
+    b = rand(9)
+
+    df = (y = y, a = 1:9, b = b, c = repeat(["d", "e", "f"], 3))
     f = apply_schema(f, schema(f, df))
     @test f == apply_schema(f, schema(f, df))
 
+    df2 = (y = y, a = 1:9, b = b, c = repeat(["d", "e", "f"], 3))
+
+    @test schema(df) == schema(df2)
+    @test isequal(schema(df), schema(df2))
```
Comment (kleinschmidt) [inline test/schema.jl]: this is a good start but you should also test as many of the "edge cases" you can think of as possible, since this will serve to make sure we don't break things in the future.  In particular, I'd suggest adding tests for 

* schema/terms with different named variables are not equal
* schema/terms with different invariants (mean/var/min/max, contrasts) are not equal
* schema are generated based on different tables with the same invariants (e.g., df1 with each column repeated twice) are equal

And whatever else you can dream up :)

**Comment by mrrobot-2000** [comment]
Hi, Thanks for reviewing. I'll do the changes and add a summary of expected behaviour.

File: test/schema.jl (hunk 2 above)
Comment (mrrobot-2000) [reply test/schema.jl]: I understood the first two points. Can you explain the last one in some detail ?

Comment (kleinschmidt) [reply test/schema.jl]: the schema shouldn't depend on the exact data that generated it, as long as the unique levels (for categorical) or mean/std/min/max (for continuous) are the same.  you check it right now by creating what is essentially a deep copy of the table, but it would be good to also check using a table that's not exactly the same size but has the same invariants, e.g. something like

`df2 = (y = [df1.y; df1.y], a = [1:9; 1:9], b = [b; b], c = [df.c; df.c])`

File: src/schema.jl [hunk 3]
```diff
@@ -53,6 +53,29 @@ Base.merge!(a::Schema, b::Schema) = (merge!(a.schema, b.schema); a)
 Base.keys(schema::Schema) = keys(schema.schema)
 Base.haskey(schema::Schema, key) = haskey(schema.schema, key)
 
+function ==(first::Schema, second::Schema)
+    first === second && return true
+    first.schema === second.schema && return true
+    length(first.schema) != length(second.schema) && return false
+    for key in keys(first)
+        !haskey(second, key) && return false
+        get(second, key, nothing) != get(first, key, nothing) && return false
```
Comment (kleinschmidt) [inline src/schema.jl]: I don't think `get` is necessary here since you've already checked that the key is present on both sides.
```suggestion
        second[key] != first[key] && return false
```

File: src/schema.jl [hunk 4]
```diff
@@ -53,6 +53,29 @@ Base.merge!(a::Schema, b::Schema) = (merge!(a.schema, b.schema); a)
...
+end
+
+function Base.isequal(first::Schema, second::Schema)
+    first === second && return true
+    first.schema === second.schema && return true
+    length(first.schema) != length(second.schema) && return false
+    for key in keys(first)
+        !haskey(second, key) && return false
+        !isequal(get(second, key, nothing), get(first, key, nothing)) &&
+        return false
```
Comment (kleinschmidt) [inline src/schema.jl]: Same here

```suggestion
        !isequal(second[key], first[key]) && return false
```

File: src/terms.jl [hunk 5]
```diff
@@ -250,6 +286,13 @@ end
 MatrixTerm(t::AbstractTerm) = MatrixTerm((t, ))
 width(t::MatrixTerm) = sum(width(tt) for tt in t.terms)
 
+# ==(first::MatrixTerm, second::MatrixTerm) =
+#    collect_matrix_terms(first.terms) == collect_matrix_terms(second.terms)
+# isequal(first::MatrixTerm, second::MatrixTerm) =
+#    isequal(collect_matrix_terms(first.terms), collect_matrix_terms(second.terms))
```
Comment (kleinschmidt) [inline src/terms.jl]: explain why these are still here or delete

File: test/schema.jl [hunk 6]
```diff
@@ -1,10 +1,39 @@
 @testset "schemas" begin
-
+    import Base
```
Comment (kleinschmidt) [inline test/schema.jl]: is this needed?

File: test/schema.jl [hunk 7]
```diff
@@ -1,10 +1,39 @@
...
+    @test isequal(schema(df), schema(df2))
+    @test !isequal(schema(df), schema(df3))
+    @test !isequal(schema(df), schema(df4))
+    @test !isequal(schema(df), schema(df5))
+    @test !isequal(sch, sch2)
+
+    # @test schema(df) == schema(df3)
+    # @test isequal(schema(df), schema(df3))
+    #@test schema(df) != schema(df4)
+    #@test isequal(schema(df), schema(df4))
```
Comment (kleinschmidt) [inline test/schema.jl]: remove

File: test/schema.jl [hunk 8]
```diff
@@ -1,10 +1,39 @@
...
+    b = rand(9)
+
+    df = (y = y, a = 1:9, b = b, c = repeat(["d", "e", "f"], 3))
     f = apply_schema(f, schema(f, df))
     @test f == apply_schema(f, schema(f, df))
 
+    df2 = (y = y, a = 1:9, b = b, c = [df.c; df.c])
+    df3 = (y = y, a = 1:9, b = b, c = repeat(["a", "b", "c"], 3))
+    df4 = (y = [df.y; df.y], a = [1:9; 1:9], b = [b; b], c = [df.c; df.c])
+    df5 = (z = y, a = 1:9, b = b, c = repeat(["d", "e", "f"], 3))
```
Comment (kleinschmidt) [inline test/schema.jl]: A few more cases to test: same values but different keys (both more/fewer keys in first/second, and equal numbers but different keys).

File: src/terms.jl [hunk 9]
```diff
@@ -211,6 +233,14 @@ struct ContinuousTerm{T} <: AbstractTerm
 end
 width(::ContinuousTerm) = 1
 
+==(first::ContinuousTerm, second::ContinuousTerm) = first.sym == second.sym &&
```
Comment (kleinschmidt) [inline src/terms.jl]: consistent style:

```suggestion
==(first::ContinuousTerm, second::ContinuousTerm) = 
    first.sym == second.sym &&
```

File: src/terms.jl [hunk 10]
```diff
@@ -187,6 +205,10 @@ via the [`implicit_intercept`](@ref) trait).
 struct InterceptTerm{HasIntercept} <: AbstractTerm end
 width(::InterceptTerm{H}) where {H} = H ? 1 : 0
 
+==(first::InterceptTerm, second::InterceptTerm) = width(first) == width(second)
```
Comment (kleinschmidt) [inline src/terms.jl]: consistent style
```suggestion
==(first::InterceptTerm, second::InterceptTerm) = 
    width(first) == width(second)
```

File: src/terms.jl [hunk 11]
```diff
@@ -127,6 +135,12 @@ FunctionTerm(forig::Fo, fanon::Fa, names::NTuple{N,Symbol},
     FunctionTerm{Fo, Fa, names}(forig, fanon, exorig, args_parsed)
 width(::FunctionTerm) = 1
 
+==(first::FunctionTerm, second::FunctionTerm) = first.forig == second.forig &&
```
Comment (kleinschmidt) [inline src/terms.jl]:
```suggestion
==(first::FunctionTerm, second::FunctionTerm) =
    first.forig == second.forig &&
```

File: src/terms.jl [hunk 12]
```diff
@@ -54,6 +57,11 @@ struct FormulaTerm{L,R} <: AbstractTerm
     rhs::R
 end
 
+==(first::FormulaTerm, second::FormulaTerm) =
+    first.lhs == second.lhs && first.rhs == second.rhs
+isequal(first::FormulaTerm, second::FormulaTerm) = isequal(first.lhs, second.lhs) &&
```
Comment (kleinschmidt) [inline src/terms.jl]:
```suggestion
isequal(first::FormulaTerm, second::FormulaTerm) =
    isequal(first.lhs, second.lhs) &&
```

**Comment by mrrobot-2000** [comment]
I have fixed all things you asked except adding tests for Matrix terms. I do not know how to create a matrix term or how to include one in a formula. Can you please give an example of one such term ?

**Comment by kleinschmidt** [comment]
They come from doing `apply_schema`, which creates them on the RHS of the formula.

**Comment by kleinschmidt** [comment]
New tests still not passing.

**Comment by mrrobot-2000** [comment]
Hi! I have fixed the broken tests. The latest build is now passing.

File: test/schema.jl [hunk 13]
```diff
@@ -1,10 +1,47 @@
...
+    @test schema(df) == schema(df2)
+    @test apply_schema(f, schema(df)) == apply_schema(f, schema(df2))
+    @test schema(df) != schema(df3)
+    @test schema(df) != schema(df4)
+    @test schema(df) != schema(df5)
+    @test schema(df) != schema(df6)
+    @test schema(df) != schema(df7)
+    @test schema(df) != schema(df8)
+    @test schema(df8) != schema(df)
+    @test apply_schema(f, schema(df)) == apply_schema(f, schema(df5))
```
Comment (kleinschmidt) [inline test/schema.jl]: First I was confused because `schema(df) != schema(df5)`, so the result of applying them to the same formula shouldn't be equal. But then I realized that `f` has already had a schema applied to it above, so this isn't testing what I thought it was testing.  Which makes me wonder, what is the purpose of this test?  `apply_schema(@formula(y ~ 1 + a + log(b) + c + b&c), schema(df5))` would produce an error because `y` isn't found in `df5`.  But since you've already applied a schema with `y` in it above, you're actually just re-applying the schema terms that ARE found. I'm not sure that's a particularly useful test, unless you were specifically trying to test this edge case... so can you explain what the goal of this test is?
